src/ucp/core/ucp_worker.c

```diff
@@ -2350,3 +2361,94 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
 
     UCP_WORKER_THREAD_CS_EXIT_CONDITIONAL(worker);
 }
+
+static int ucp_worker_keepalive_is_enabled(ucp_worker_h worker)
+{
+    return (worker->context->config.ext.keepalive_timeout != 0) &&
+           (worker->context->config.ext.keepalive_num_eps != 0);
+}
+
+static unsigned ucp_worker_keepalive_cb(void *arg)
+{
+    ucp_worker_h worker = (ucp_worker_h)arg;
+    ucs_time_t now      = ucs_get_time();
+    unsigned cnt;
+    ucp_ep_ext_gen_t *ep_ext;
+    ucs_list_link_t *iter_begin;
+
+    if (ucs_likely((now - worker->keepalive.last_round) <
+                   worker->context->config.ext.keepalive_timeout)) {
+        return 0;
+    }
+
+    worker->keepalive.last_round = now;
+    cnt                          = 0;
+    iter_begin                   = worker->keepalive.iter;
+    /* use own loop for elements because standard for_each skips
+     * head element */
+    /* TODO: use more optimal algo to enumerate EPs to keepalive
+     * (linked list) */
+    do {
+        ep_ext = ucs_list_next(worker->keepalive.iter,
+                               ucp_ep_ext_gen_t, ep_list);
+        worker->keepalive.iter = &ep_ext->ep_list;
+        if (worker->keepalive.iter == &worker->all_eps) {
+            continue;
+        }
+
+        cnt += ucp_ep_do_keepalive(ucp_ep_from_ext_gen(ep_ext));
+    } while ((iter_begin != &ep_ext->ep_list) &&
+             (cnt < worker->context->config.ext.keepalive_num_eps));
+
+    if (cnt == 0) {
+        /* nothing to keepalive? unsubscribe */
+        uct_worker_progress_unregister_safe(worker->uct,
+                                            &worker->keepalive.cb_id);
+    }
+
+    return cnt;
+}
+
+void ucp_worker_keepalive_add_ep(ucp_worker_h worker)
+{
+    if (!ucp_worker_keepalive_is_enabled(worker)) {
+        return;
+    }
+
+    uct_worker_progress_register_safe(worker->uct, ucp_worker_keepalive_cb,
+                                      worker, 0, &worker->keepalive.cb_id);
+}
+
+/* EP is removed from worker */
+void ucp_worker_keepalive_remove_ep(ucp_worker_h worker, ucp_ep_h remove_ep)
+{
+    ucp_ep_ext_gen_t *ep_ext;
+    ucp_ep_h ep;
+
+    if (!ucp_worker_keepalive_is_enabled(worker)) {
+        ucs_assert(worker->keepalive.iter == &worker->all_eps);
+        return;
+    }
+
+    /* EP is removed from worker all_eps, so we can't use list->next value */
+    if (worker->keepalive.iter == &ucp_ep_ext_gen(remove_ep)->ep_list) {
+        ep_ext = ucs_list_next(worker->keepalive.iter,
+                               ucp_ep_ext_gen_t, ep_list);
+        worker->keepalive.iter = &ep_ext->ep_list;
+    }
+
+    if (worker->keepalive.cb_id == UCS_CALLBACKQ_ID_NULL) {
+        return; /* already unsubscribed */
+    }
+
+    /* look for first keepalive EP  not same as removed EP & return */
+    ucs_list_for_each(ep_ext, &worker->all_eps, ep_list) {
+        ep = ucp_ep_from_ext_gen(ep_ext);
+        if ((ep != remove_ep) && ucp_ep_keepalive_is_enabled(ep)) {
+            return;
+        }
+    }
```

Comment:
1. remove this
2. add cbq remove in worker_destroy
3. the ka progress remove itself if no ka eps exist

Reply:
added unsubscribe on worker destroy and removed from ucp_worker_keepalive_remove_ep